src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs

```diff
@@ -0,0 +1,32 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GraphQL.DataLoader;
+using GraphQL.Types;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Apis.GraphQL;
+using OrchardCore.ContentManagement.Records;
+using YesSql;
+using YesSql.Services;
+
+namespace OrchardCore.ContentManagement.GraphQL
+{
+    public static class DataLoaderExtensions
+    {
+        public static IDataLoader<string, ContentItem> GetorAddPublishedContentItemByIdDataLoader<T>(this ResolveFieldContext<T> context)
+        {
+            var serviceProvider = context.ResolveServiceProvider();
+
+            var accessor = serviceProvider.GetRequiredService<IDataLoaderContextAccessor>();
+            var session = serviceProvider.GetService<ISession>();
+
+            return accessor.Context.GetOrAddBatchLoader<string, ContentItem>("GetPublishedContentItemsById", ci => LoadPublishedContentItems(ci, session));
+        }
+
+        private static async Task<IDictionary<string, ContentItem>> LoadPublishedContentItems(IEnumerable<string> contentItemIds, ISession session)
+        {
+            var contentItemsLoaded = await session.Query<ContentItem, ContentItemIndex>(y => y.ContentItemId.IsIn(contentItemIds) && y.Published).ListAsync();
```

Comment:
It needs to be ordered with the same order as the key, you lost this piece:
```
contentItems.OrderBy(c => Array.IndexOf(x.Source.ContentItemIds, c.ContentItemId));
```

Reply:
The dataloader does this by default when it passes the keys to loadasync (maintains the order of the passed keys)

Reply:
awesome, thanks